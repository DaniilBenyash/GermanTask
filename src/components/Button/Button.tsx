import { DetailedHTMLProps, FC, HTMLAttributes, PropsWithChildren, ComponentProps } from "react";
import styles from "./Button.module.scss";
import cn from "classnames";

interface ButtonProps extends DetailedHTMLProps<React.ButtonHTMLAttributes<HTMLButtonElement>, HTMLButtonElement> {
  variant?: "default" | "primary" | "secondary" | "default_mobile";
}

export const Button: FC<ButtonProps> = ({ variant = "default", className, ...props }) => {
  const combinedClassName = cn(styles.button, styles[`button_${variant}`], className);
  return (
    <button {...props} className={combinedClassName} />
  );
};
