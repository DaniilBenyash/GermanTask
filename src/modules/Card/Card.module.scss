.card {
  display: flex;
  flex-direction: column;
  gap: 8px;
  padding: 8px 16px;
  background-color: #ffffff;
  border: #707070 1px solid;
  color: #707070;
  cursor: pointer;

  &:active {
    opacity: 0.8;
  }
}

.title {
  font-size: 16px;
}
.description {
  font-size: 16px;
  overflow: hidden;
  white-space: nowrap;
}
@media screen and (max-width: 500px) {
  .card {
    border-top: none;
  }
}
