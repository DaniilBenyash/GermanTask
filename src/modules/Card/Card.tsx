import { FC } from "react";
import styles from "./Card.module.scss";

type CardProps = {
  title: string;
  email: string;
  onClick: () => void;
};
export const Card: FC<CardProps> = ({ title, email, onClick }) => {
  return (
    <button className={styles.card} onClick={onClick}>
      <h3 className={styles.title}>{title}</h3>
      <p className={styles.description}>{email}</p>
    </button>
  );
};
