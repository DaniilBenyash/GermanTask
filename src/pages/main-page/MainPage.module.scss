.main_page {
  min-height: 100vh;
  background-color: #ebecec;
}
.section {
  display: flex;
  flex-direction: column;
  align-items: flex-start;
  justify-content: center;
  gap: 39px;
  padding: 48px;
}
.cards_section {
  display: grid;
  gap: 20px;
  grid-template-columns: repeat(auto-fit, minmax(200px, 1fr));
}

.mobile_button {
  display: none;
}
.box_contact {
  width: 100%;
}
@media screen and (max-width: 500px) {
  .section {
    gap: 0;
    padding: 0;
  }

  .cards_section {
    width: 100%;
    gap: 0;
    grid-template-columns: 1fr;
  }
  .desktop_button {
    display: none;
  }
  .mobile_button {
    display: block;
    position: fixed;
    bottom: 30px;
    right: 30px;
  }
}
