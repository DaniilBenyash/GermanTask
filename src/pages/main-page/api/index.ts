import { createAPI } from "../../helpers/createApi";
import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";

export interface Contact {
  id: number;
  firstname: string;
  lastname: string;
  email: string;
}

const contactsKeys = {
  contacts: ["contacts"],
};

const contactsApi = createAPI("/contacts");

export const useGetUsers = () => {
  return useQuery({
    queryKey: contactsKeys.contacts,
    queryFn: async () => {
      const response = await contactsApi.get("/");
      return response.data.responseObject as Contact[];
    },
  });
};

export const useAddUser = () => {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: (contact: Contact) => contactsApi.post("/", contact),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: contactsKeys.contacts });
    },
  });
};

export const useUpdateUser = () => {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: (updatedUser: Contact) => contactsApi.put(`/${updatedUser.id}`, updatedUser),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: contactsKeys.contacts });
    },
  });
};

export const useDeleteUser = () => {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: (id: number) => contactsApi.delete(`/${id}`),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: contactsKeys.contacts });
    },
  });
};
