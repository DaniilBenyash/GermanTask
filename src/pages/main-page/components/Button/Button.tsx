import { FC } from "react";
import styles from "./Button.module.scss";
import { Button as ButtonUI } from "../../../../ui-kit";

type ButtonProps = {
  type: "desktop" | "mobile";
  onClick: () => void;
};

export const Button: FC<ButtonProps> = ({ type, onClick }) => {
  if (type === "desktop") {
    return (
      <div className={styles.desktop_button}>
        <ButtonUI type="default" onClick={onClick}>
          NEUER EINTRAG
        </ButtonUI>
      </div>
    );
  }
  if (type === "mobile") {
    return (
      <div className={styles.mobile_button}>
        <ButtonUI type="default_mobile" onClick={onClick}>
          +
        </ButtonUI>
      </div>
    );
  }
};
