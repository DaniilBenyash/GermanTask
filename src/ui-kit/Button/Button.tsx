import { FC, ReactNode } from "react";
import styles from "./Button.module.scss";
import cn from "classnames";

type ButtonProps = {
  type?: "default" | "primary" | "secondary" | "default_mobile";
  children: ReactNode;
  onClick?: () => void;
  disabled?: boolean;
};

export const Button: FC<ButtonProps> = ({ type = "default", onClick, children, disabled }) => {
  const className = cn(styles.button, {
    [styles.button_default]: type === "default",
    [styles.button_default_mobile]: type === "default_mobile",
    [styles.button_primary]: type === "primary",
    [styles.button_secondary]: type === "secondary",
  });
  return (
    <button className={className} onClick={onClick} disabled={disabled}>
      {children}
    </button>
  );
};
