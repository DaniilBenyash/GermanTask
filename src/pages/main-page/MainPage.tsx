import styles from "./MainPage.module.scss";
import { Header, Card, Modal, Button } from "../../components";
import { Form } from "./components/Form";
import { useState } from "react";
import { useAddUser, useDeleteUser, useGetUsers, useUpdateUser } from "./api";
import { Contact } from "../../types/types";

export const MainPage = () => {
  const { data } = useGetUsers();
  const { mutate: onAddUser } = useAddUser();
  const { mutate: onUpdateUser } = useUpdateUser();
  const { mutate: onDeleteUser } = useDeleteUser();
  const [isOpenModal, setIsOpenModal] = useState(false);
  const [contact, setContact] = useState<Contact | null>(null);

  const handleCloseModal = () => {
    setContact(null);
    setIsOpenModal(false);
  };
  const handleOpenModal = () => {
    setIsOpenModal(true);
  };
  const handleOpenModalWithData = (contact: Contact) => {
    setContact(contact);
    setIsOpenModal(true);
  };

  const isMobile = window.innerWidth < 534;

  return (
    <main className={styles.main_page}>
      <Header />
      <section className={styles.section}>
        <Button className={isMobile ? styles.mobile_button : styles.desktop_button} onClick={handleOpenModal}>
          {isMobile ? <p className={styles.plus}>+</p> : 'NEUER EINTRAG'}
        </Button>
        <div className={styles.cards_section}>
          {data?.map((contact) => {
            return (
              <Card
                key={contact.id}
                title={`${contact.firstname} ${contact.lastname}`}
                email={contact.email}
                onClick={() => handleOpenModalWithData(contact)}
              />
            );
          })}
        </div>
      </section>
      <Modal isOpen={isOpenModal} onClose={handleCloseModal}>
        <Form
          contact={contact}
          onUpdateUser={onUpdateUser}
          onAddUser={onAddUser}
          onDeleteUser={onDeleteUser}
          onCloseModal={handleCloseModal}
        />
      </Modal>
    </main>
  );
};
