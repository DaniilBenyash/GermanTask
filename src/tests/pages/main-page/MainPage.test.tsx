import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import { MainPage } from "../../../pages/main-page";
import renderer from "react-test-renderer";
import { render, screen, fireEvent } from "@testing-library/react";
import { CardProps } from "../../../modules/Card";
import { ButtonProps } from "../../../pages/main-page/components/Button";
import { ModalProps } from "../../../modules/Modal";

const queryClient = new QueryClient();

it("renders correctly", () => {
  const tree = renderer
    .create(
      <QueryClientProvider client={queryClient}>
        <MainPage />
      </QueryClientProvider>,
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});


jest.mock("../../../pages/main-page/api", () => ({
  useGetUsers: jest.fn(() => ({
    isLoading: false,
    data: [{ id: "1", firstname: "A", lastname: "B", email: "AB@gmail.com" }],
  })),
  useAddUser: jest.fn(() => ({ mutate: jest.fn() })),
  useUpdateUser: jest.fn(() => ({ mutate: jest.fn() })),
  useDeleteUser: jest.fn(() => ({ mutate: jest.fn() })),
}));

jest.mock("../../../modules/Header", () => ({
  Header: () => <div>Header Mock</div>,
}));

jest.mock("../../../modules/Card", () => ({
  Card: ({ onClick, title }: CardProps) => <div onClick={onClick}>{title}</div>,
}));

jest.mock("../../../pages/main-page/components/Button", () => ({
  Button: ({ onClick }: ButtonProps) => <button onClick={onClick}>Open Modal</button>,
}));

jest.mock("../../../modules/Modal", () => ({
  Modal: ({ isOpen, children }: ModalProps) =>
    isOpen ? <div>{children}</div> : null,
}));

jest.mock("../../../pages/main-page/components/Form", () => ({
  Form: () => <div>Form Content</div>,
}));


jest.mock("../../../modules/Modal", () => ({
  Modal: ({ isOpen, onClose, children }: ModalProps) => isOpen ? (
    <div onClick={event => {
      if (event.target === event.currentTarget) onClose();
    }} data-testid="modal-backdrop">
      {children}
    </div>
  ) : null
}));

describe('MainPage Modal', () => {
  it('open modal without data', () => {
    render(<MainPage />);
    fireEvent.click(screen.getAllByText('Open Modal')[0]);
    expect(screen.getByTestId('modal-backdrop')).toBeInTheDocument();
  });

  it('close modal', () => {
    render(<MainPage />);
    fireEvent.click(screen.getAllByText('Open Modal')[0]);
    const modalBackdrop = screen.getByTestId('modal-backdrop');
    fireEvent.click(modalBackdrop);
    expect(screen.queryByTestId('modal-backdrop')).not.toBeInTheDocument();
  });

  it('open modal', () => {
    render(<MainPage />);
    fireEvent.click(screen.getByText('A B'));
    expect(screen.getByTestId('modal-backdrop')).toBeInTheDocument();
    expect(screen.getByText('Form Content')).toBeInTheDocument(); 
  });
});
