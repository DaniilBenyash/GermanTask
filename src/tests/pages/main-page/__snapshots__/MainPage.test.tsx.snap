// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<main
  className="main_page"
>
  <div>
    Header Mock
  </div>
  <section
    className="section"
  >
    <button
      onClick={[Function]}
    >
      Open Modal
    </button>
    <button
      onClick={[Function]}
    >
      Open Modal
    </button>
    <div
      className="cards_section"
    >
      <div
        onClick={[Function]}
      >
        A B
      </div>
    </div>
  </section>
</main>
`;
